### PR TITLE
Llin/ent 6148 fullfill exec ed

### PR DESCRIPTION
**Pull request by long74100**
https://2u-internal.atlassian.net/browse/ENT-6148

# ⛔️ DEPRECATION WARNING

**This repository is deprecated and in maintainence-only operation while we work on a replacement, please see [this announcement](https://discuss.openedx.org/t/deprecation-removal-ecommerce-service-depr-22/6839) for more information.**

Although we have stopped integrating new contributions, we always appreciate security disclosures and patches sent to [security@edx.org](mailto:security@edx.org)

## Required Testing
- [ ] Before deploying this change, complete a purchase in the stage environment. 
(^ We can remove that manual check once REV-2624 is done and the corresponding e2e test runs again)

## Description

Describe what this pull request changes, and why these changes were made. How will these changes affect other people, installations of edx, etc.?
Please include links to any relevant ADRs, design artifacts, and decision documents. Make sure to document the rationale behind significant changes in the repo, per [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be
linked here.

Useful information to include:
- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author", "Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these changes.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

Please provide detailed step-by-step instructions for testing this change; how did YOU test this change?

## Other information

Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, OpenEdx vs. edx.org differences, development vs. production environment differences, security, or accessibility.

### CODE REVIEW

File: ecommerce/extensions/executive_education_2u/views.py [hunk 1]
```diff
@@ -168,7 +168,14 @@ def finish_checkout(self, request):
 
                 return Response('Failed to create subsidized order.', status=status.HTTP_422_UNPROCESSABLE_ENTITY)
 
-            order = self.place_free_order(basket)
+            order = self.place_free_order(
+                basket=basket,
+                address=request.data['address'],
```
Comment (adamstankiewicz) [inline ecommerce/extensions/executive_education_2u/views.py:175]: [curious] Based on this implementation, the `POST` call's payload will provide a singular address field from the client. Is the recommendation here that we'd concatenate the Address Line 1 + Address Line 2 fields together from the Exec Ed (2U) terms/profile form in the UI? Does this address data also intend to include city/state/country, or are those provided via another means?

File: ecommerce/extensions/fulfillment/tests/test_modules.py [hunk 2]
```diff
@@ -1152,3 +1154,125 @@ def test_entitlement_module_fulfill_bad_attributes(self):
...
+            'Regular Course Entitlement'
+        )
+
+        basket = factories.BasketFactory(owner=self.user, site=self.site)
+        basket.add_product(self.exec_ed_2u_course_entitlement, 1)
+        basket.add_product(self.exec_ed_2u_course_entitlement_2, 1)
+        basket.add_product(self.non_exec_ed_2u_course_entitlement, 1)
+        self.order = create_order(number=1, basket=basket, user=self.user)
+
+        order_lines =  self.order.lines.all()
```
Comment (iloveagent57) [inline ecommerce/extensions/fulfillment/tests/test_modules.py]: nit: extra space to right of `=`

File: ecommerce/extensions/executive_education_2u/views.py [hunk 3]
```diff
@@ -168,7 +168,14 @@ def finish_checkout(self, request):
...
+                user_details={**request.data['user_details'], 'email': request.user.email},
```
Comment (adamstankiewicz) [inline ecommerce/extensions/executive_education_2u/views.py:176]: [clarification] What would the expected contents of `user_details` data be when passing it in the `POST` payload? Are these required data documented somewhere?

File: ecommerce/extensions/fulfillment/modules.py [hunk 4]
```diff
@@ -902,3 +904,151 @@ def revoke_line(self, line):
...
+        Args:
+            lines (List of Lines): Order Lines, associated with purchased products in an Order.
+
+        Returns:
+            A supported list of lines, unmodified.
+        """
+        return [line for line in lines if self.supports_line(line)]
+
+    def _generate_payment_reference(self):
+        # TODO: Implemen method for generating payment reference
```
Comment (iloveagent57) [inline ecommerce/extensions/fulfillment/modules.py]: referring to this, right? https://2u-internal.atlassian.net/browse/ENT-6088

Comment (adamstankiewicz) [inline ecommerce/extensions/fulfillment/modules.py]: nit: typo "Implement"

File: ecommerce/extensions/fulfillment/modules.py [hunk 5]
```diff
@@ -902,3 +904,151 @@ def revoke_line(self, line):
...
+
+            allocation_payload = self._create_allocation_payload(
+                line=line,
+                fulfillment_details=fulfillment_details
+            )
+
+            try:
+                self.get_smarter_client.create_allocation(**allocation_payload)
+            except Exception:
+                logger.exception("Fulfillment of line [%d] on order [%s] failed.", line.id, order.number)
```
Comment (iloveagent57) [inline ecommerce/extensions/fulfillment/modules.py]: nit: it'd be nice to say why it failed (not sure how much the GS client returns in failure payloads).

File: ecommerce/extensions/fulfillment/modules.py [hunk 6]
```diff
@@ -902,3 +904,151 @@ def revoke_line(self, line):
...
+                as part of the fulfillment. Defaults to False.
+
+        Returns:
+            The original set of lines, with new statuses set based on the success or failure of fulfillment.
+
+        """
+
+        lines = [line for line in lines if self.supports_line(line)]
+
+        logger.info("Attempting to fulfill Executive Education 2U product for order [%s]", order.number)
```
Comment (adamstankiewicz) [inline ecommerce/extensions/fulfillment/modules.py]: nit: "Executive Education (2U)"

**Review by iloveagent57 — APPROVED** [review]
Looks good to me.  One small suggestion: let's get as much info from the GS endpoint response object as we can when the request fails.  It likely has _something_ jammed into either `response.text` or `response.json()`.

File: ecommerce/extensions/executive_education_2u/views.py (hunk 1 above)
Comment (long74100) [reply ecommerce/extensions/executive_education_2u/views.py:175]: This is the serializer that validates the payload right now https://github.com/openedx/ecommerce/blob/master/ecommerce/extensions/executive_education_2u/serializers.py. It does not have address2 right now, but we can extend this to accommodate all the other fields.

Comment (adamstankiewicz) [reply ecommerce/extensions/executive_education_2u/views.py:175]: Ah got it, thank you!

Comment (adamstankiewicz) [reply ecommerce/extensions/executive_education_2u/views.py:175]: Yeah, we'll probably want to extend to include `address_line2` under the `AddressSerializer` along with `work_experience` and `education_level` as those are also (optional) fields included in the UI. Separate task?

Comment (long74100) [reply ecommerce/extensions/executive_education_2u/views.py:175]: I can add it here.

Comment (long74100) [reply ecommerce/extensions/executive_education_2u/views.py:175]: added

**Review by iloveagent57 — APPROVED** [review]

**Review by adamstankiewicz — APPROVED** [review]

**Pull request merged**